File: RaWin/ViewModels/CommandsPanelViewModel.cs
using System.ComponentModel;
using System.Runtime.CompilerServices;
using System.Windows.Input;

namespace RaWin.ViewModels
{
    /// <summary>
    /// ViewModel for the Commands card in Monitoring.
    /// </summary>
    public class CommandsPanelViewModel : INotifyPropertyChanged
    {
        private string _commandInput;
        public string CommandInput
        {
            get => _commandInput;
            set { _commandInput = value; OnPropertyChanged(); }
        }

        private string _commandResult;
        public string CommandResult
        {
            get => _commandResult;
            set { _commandResult = value; OnPropertyChanged(); }
        }

        private string _errorMessage;
        public string ErrorMessage
        {
            get => _errorMessage;
            set { _errorMessage = value; OnPropertyChanged(); }
        }

        public ICommand SendCommand { get; }

        public CommandsPanelViewModel()
        {
            SendCommand = new RelayCommand(_ => Execute());
        }

        private void Execute()
        {
            if (string.IsNullOrWhiteSpace(CommandInput))
            {
                ErrorMessage = "Command cannot be empty.";
                CommandResult = "";
            }
            else
            {
                // Placeholder for actual command execution logic. Extend as needed.
                CommandResult = $"Executed: {CommandInput}";
                ErrorMessage = "";
            }
            OnPropertyChanged(nameof(CommandResult));
            OnPropertyChanged(nameof(ErrorMessage));
        }

        public event PropertyChangedEventHandler PropertyChanged;
        protected void OnPropertyChanged([CallerMemberName] string prop = null)
            => PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(prop));
    }
}

File: RaStudios/viewmodels/commands_panel_viewmodel.py
from viewmodels.observable_object import ObservableObject
from viewmodels.relay_command import RelayCommand

class CommandsPanelViewModel(ObservableObject):
    """
    ViewModel for the Commands card in Monitoring.
    Wired to SpeechPipelineService for async agentic command execution.
    """
    def __init__(self, speech_pipeline, module_bus):
        super().__init__()
        self.speech_pipeline = speech_pipeline
        self.module_bus = module_bus

        self.command_input = ""
        self.command_result = ""
        self.error_message = ""

        self.send_command = RelayCommand(self.execute)

    def execute(self, _=None):
        if not self.command_input.strip():
            self.error_message = "Command cannot be empty."
            self.command_result = ""
        else:
            try:
                self.command_result = self.speech_pipeline.send_async(self.command_input)
                self.error_message = ""
            except Exception as ex:
                self.error_message = f"Error: {type(ex).__name__}: {ex}"
                self.command_result = ""
                self.module_bus.publish("Diagnostics", {
                    "timestamp": "now",
                    "message": self.error_message,
                    "severity": "Error",
                    "source": "CommandsPanelViewModel",
                    "code": "COMMAND_ERROR",
                    "metadata": {"CommandInput": self.command_input}
                })
        self.notify_property_changed("command_result")
        self.notify_property_changed("error_message")